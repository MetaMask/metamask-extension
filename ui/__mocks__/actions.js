const actions = require('../store/actions');
const {
  CONTRACT_ADDRESS_UNISWAP,
  CONTRACT_ADDRESS_SOURCIFY,
  CONTRACT_ADDRESS_FOUR_BYTE,
  TRANSACTION_DECODE_UNISWAP,
  TRANSACTION_DECODE_SOURCIFY,
  TRANSACTION_DECODE_FOUR_BYTE,
} = require('../../test/data/confirmations/transaction-decode');

const ERC20_TOKEN_1_MOCK = '0x2260fac5e5542a773aa44fbcfedf7c193bc2c599'; // WBTC
const ERC20_TOKEN_2_MOCK = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'; // USDC

const TOKEN_DETAILS_MOCK = {
  [ERC20_TOKEN_1_MOCK]: {
    address: ERC20_TOKEN_1_MOCK,
    standard: 'ERC20',
    decimals: 8,
    name: 'Wrapped Bitcoin',
  },
  [ERC20_TOKEN_2_MOCK]: {
    address: ERC20_TOKEN_2_MOCK,
    standard: 'ERC20',
    decimals: 6,
    name: 'USD Coin',
  },
};

module.exports = {
  ...actions,
  getTokenStandardAndDetails: (address) => {
    if (!TOKEN_DETAILS_MOCK[address]) {
      console.log('Token not found:', address);

      return Promise.resolve({
        address,
        standard: 'ERC20',
        decimals: 18,
        name: 'Missing Mock',
      });
    }

    return Promise.resolve(TOKEN_DETAILS_MOCK[address]);
  },

  // eslint-disable-next-line no-empty-function
  trackMetaMetricsEvent: () => {},

  decodeTransactionData: async (request) => {
    const { contractAddress } = request;

    if (contractAddress === CONTRACT_ADDRESS_UNISWAP) {
      return TRANSACTION_DECODE_UNISWAP;
    } else if (contractAddress === CONTRACT_ADDRESS_SOURCIFY) {
      return TRANSACTION_DECODE_SOURCIFY;
    } else if (contractAddress === CONTRACT_ADDRESS_FOUR_BYTE) {
      return TRANSACTION_DECODE_FOUR_BYTE;
    }

    return undefined;
  },
};
