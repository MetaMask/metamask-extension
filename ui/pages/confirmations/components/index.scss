/** Please import your files in alphabetical order **/
@import 'advanced-gas-controls/index';
@import 'advanced-gas-fee-popover/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-gas-limit/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-defaults/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-inputs/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-inputs/base-fee-input/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-inputs/priority-fee-input/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-input-subtext/index';
@import 'approve-content-card/index';
@import 'confirmation-warning-modal/index';
@import 'confirm-page-container/index';
@import 'confirm/footer';
@import 'confirm/header/header.scss';
@import 'confirm/scroll-to-bottom';
@import 'confirm/nav/nav.scss';
@import "confirm/network-change-toast/index";
@import 'contract-details-modal/index';
@import 'custom-nonce/index';
@import 'edit-gas-display/index';
@import 'edit-gas-fee-button/index';
@import 'edit-gas-fee-popover/index';
@import 'edit-gas-fee-popover/edit-gas-item/index';
@import 'edit-gas-fee-popover/network-statistics/index';
@import 'edit-gas-fee-popover/network-statistics/status-slider/index';
@import 'edit-gas-fee-popover/edit-gas-tooltip/index';
@import 'edit-gas-popover/index';
@import 'gas-details-item/index';
@import 'multilayer-fee-message/index';
@import 'review-spending-cap/index';
@import 'set-approval-for-all-warning/index';
@import 'signature-request/index';
@import 'signature-request-original/index';
@import 'signature-request-siwe/index';
@import 'transaction-alerts/transaction-alerts';
@import 'transaction-decoding/index';
@import 'transaction-detail/index';
@import 'transaction-detail-item/index';
@import 'simulation-details/index';
