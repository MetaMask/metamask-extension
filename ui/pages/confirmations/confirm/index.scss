@use "design-system";

.confirm_wrapper {
  box-shadow: var(--shadow-size-lg) var(--color-shadow-default);

  @media screen and (min-width: design-system.$break-small) {
    max-width: 62%;
  }
}
