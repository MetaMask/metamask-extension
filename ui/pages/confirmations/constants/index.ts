import { CHAINLIST_CHAIN_IDS_MAP } from '../../../../shared/constants/network';

export const IGNORE_GAS_LIMIT_CHAIN_IDS = [CHAINLIST_CHAIN_IDS_MAP.MANTLE];

export const TYPED_SIGNATURE_VERSIONS = {
  V1: 'V1',
  V3: 'V3',
  V4: 'V4',
};

export const SPENDING_CAP_UNLIMITED_MSG = 'UNLIMITED MESSAGE';

export const TypedSignSignaturePrimaryTypes = {
  PERMIT: 'Permit',
  ORDER: 'Order',
};
