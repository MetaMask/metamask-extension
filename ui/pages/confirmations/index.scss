/** Please import your files in alphabetical order **/
@import 'confirm-approve/index';
@import 'confirm/index';
@import 'confirmation/confirmation';
@import 'send/gas-display/index';
@import 'send/send';
@import 'token-allowance/index';
@import './components/index'
