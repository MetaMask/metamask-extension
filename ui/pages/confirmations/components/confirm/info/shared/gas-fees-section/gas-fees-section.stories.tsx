import { Meta } from '@storybook/react';
import React from 'react';
import { Provider } from 'react-redux';
import { getMockContractInteractionConfirmState } from '../../../../../../../../test/data/confirmations/helper';
import configureStore from '../../../../../../../store/store';
import { ConfirmContextProvider } from '../../../../../context/confirm';
import { GasFeesSection } from './gas-fees-section';

function getStore() {
  return configureStore(getMockContractInteractionConfirmState());
}

const Story = {
  title: 'Components/App/Confirm/info/GasFeesSection',
  component: GasFeesSection,
  decorators: [
    (story: () => Meta<typeof GasFeesSection>) => (
      <Provider store={getStore()}>
        <div
          style={{
            backgroundColor: 'var(--color-background-alternative)',
            padding: 30,
          }}
        >
          <ConfirmContextProvider>{story()}</ConfirmContextProvider>
        </div>
      </Provider>
    ),
  ],
};

export default Story;

export const DefaultStory = () => <GasFeesSection />;

DefaultStory.storyName = 'Default';
