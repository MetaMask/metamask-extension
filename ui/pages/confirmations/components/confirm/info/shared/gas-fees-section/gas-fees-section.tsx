import { TransactionMeta } from '@metamask/transaction-controller';
import React, { useCallback, useState } from 'react';
import { EditGasModes } from '../../../../../../../../shared/constants/gas';
import { ConfirmInfoSection } from '../../../../../../../components/app/confirm/info/row/section';
import { useConfirmContext } from '../../../../../context/confirm';
import EditGasPopover from '../../../../edit-gas-popover';
import { useSupportsEIP1559 } from '../../hooks/useSupportsEIP1559';
import { GasFeesDetails } from '../gas-fees-details/gas-fees-details';

const LegacyTransactionGasModal = ({
  closeCustomizeGasPopover,
  transactionMeta,
}: {
  closeCustomizeGasPopover: () => void;
  transactionMeta: TransactionMeta;
}) => {
  return (
    <EditGasPopover
      onClose={closeCustomizeGasPopover}
      mode={EditGasModes.modifyInPlace}
      transaction={transactionMeta}
    />
  );
};

export const GasFeesSection = () => {
  const { currentConfirmation: transactionMeta } =
    useConfirmContext<TransactionMeta>();

  const [showCustomizeGasPopover, setShowCustomizeGasPopover] = useState(false);
  const closeCustomizeGasPopover = useCallback(
    () => setShowCustomizeGasPopover(false),
    [setShowCustomizeGasPopover],
  );

  const { supportsEIP1559 } = useSupportsEIP1559(transactionMeta);

  if (!transactionMeta?.txParams) {
    return null;
  }

  return (
    <ConfirmInfoSection data-testid="gas-fee-section">
      <GasFeesDetails setShowCustomizeGasPopover={setShowCustomizeGasPopover} />
      {!supportsEIP1559 && showCustomizeGasPopover && (
        <LegacyTransactionGasModal
          closeCustomizeGasPopover={closeCustomizeGasPopover}
          transactionMeta={transactionMeta}
        />
      )}
    </ConfirmInfoSection>
  );
};
