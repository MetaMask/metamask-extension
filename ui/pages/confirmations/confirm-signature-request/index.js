import PropTypes from 'prop-types';
import React, { useEffect, useState, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, withRouter } from 'react-router-dom';
import log from 'loglevel';
import { cloneDeep } from 'lodash';
import { SubjectType } from '@metamask/permission-controller';
import { TransactionStatus } from '@metamask/transaction-controller';
import * as actions from '../../../store/actions';
import txHelper from '../../../helpers/utils/tx-helper';
import SignatureRequest from '../components/signature-request';
import SignatureRequestOriginal from '../components/signature-request-original';
import SignatureRequestSIWE from '../components/signature-request-siwe';
import Loading from '../../../components/ui/loading-screen';
import { useRouting } from '../hooks/useRouting';
import {
  getTotalUnapprovedSignatureRequestCount,
  getTargetSubjectMetadata,
  getCurrentNetworkTransactions,
  getUnapprovedTransactions,
  getInternalAccounts,
  getMemoizedUnapprovedPersonalMessages,
  getMemoizedUnapprovedTypedMessages,
  getMemoizedCurrentChainId,
  getMemoizedTxId,
} from '../../../selectors';
import { useInsightSnaps } from '../../../hooks/snaps/useInsightSnaps';
import { MESSAGE_TYPE } from '../../../../shared/constants/app';
import { getSendTo } from '../../../ducks/send';

const signatureSelect = (txData, targetSubjectMetadata) => {
  const {
    type,
    msgParams: { version, siwe },
  } = txData;

  // Temporarily direct only v3 and v4 requests to new code.
  if (
    type === MESSAGE_TYPE.ETH_SIGN_TYPED_DATA &&
    (version === 'V3' || version === 'V4')
  ) {
    return SignatureRequest;
  }

  if (siwe?.isSIWEMessage && targetSubjectMetadata !== SubjectType.Snap) {
    return SignatureRequestSIWE;
  }

  return SignatureRequestOriginal;
};

const ConfirmTxScreen = ({ match }) => {
  const dispatch = useDispatch();
  const { navigateToMostRecentOverviewPage } = useRouting();
  const unapprovedMessagesTotal = useSelector(
    getTotalUnapprovedSignatureRequestCount,
  );
  const sendTo = useSelector(getSendTo);
  const internalAccounts = useSelector(getInternalAccounts);

  const { currentCurrency, blockGasLimit, signatureSecurityAlertResponses } =
    useSelector((state) => state.metamask);
  const unapprovedPersonalMsgs = useSelector(
    getMemoizedUnapprovedPersonalMessages,
  );
  const unapprovedTypedMessages = useSelector(
    getMemoizedUnapprovedTypedMessages,
  );

  const unapprovedTxs = useSelector(getUnapprovedTransactions);
  const currentNetworkTxList = useSelector(getCurrentNetworkTransactions);
  const chainId = useSelector(getMemoizedCurrentChainId);
  const index = useSelector(getMemoizedTxId);

  const [prevValue, setPrevValues] = useState();
  const history = useHistory();

  useEffect(() => {
    const unconfTxList = txHelper(
      unapprovedTxs || {},
      {},
      {},
      {},
      {},
      {},
      chainId,
    );
    if (unconfTxList.length === 0 && !sendTo && unapprovedMessagesTotal === 0) {
      navigateToMostRecentOverviewPage();
    }
  }, [
    chainId,
    navigateToMostRecentOverviewPage,
    sendTo,
    unapprovedMessagesTotal,
    unapprovedTxs,
  ]);

  useEffect(
    () => {
      if (!prevValue) {
        setPrevValues({ index, unapprovedTxs });
        return;
      }

      let prevTx;
      const { params: { id: transactionId } = {} } = match;
      if (transactionId) {
        prevTx = currentNetworkTxList.find(
          ({ id }) => `${id}` === transactionId,
        );
      } else {
        const { index: prevIndex, unapprovedTxs: prevUnapprovedTxs } =
          prevValue;
        const prevUnconfTxList = txHelper(
          prevUnapprovedTxs,
          {},
          {},
          {},
          {},
          {},
          chainId,
        );
        const prevTxData = prevUnconfTxList[prevIndex] || {};
        prevTx =
          currentNetworkTxList.find(({ id }) => id === prevTxData.id) || {};
      }

      const unconfTxList = txHelper(
        unapprovedTxs || {},
        {},
        {},
        {},
        {},
        {},
        chainId,
      );

      if (prevTx && prevTx.status === TransactionStatus.dropped) {
        dispatch(
          actions.showModal({
            name: 'TRANSACTION_CONFIRMED',
            onSubmit: () => navigateToMostRecentOverviewPage(),
          }),
        );
        return;
      }

      if (
        unconfTxList.length === 0 &&
        !sendTo &&
        unapprovedMessagesTotal === 0
      ) {
        navigateToMostRecentOverviewPage();
      }

      setPrevValues({ index, unapprovedTxs });
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [],
  );

  const { params: { id: txIdFromPath } = {} } = match;

  const txData = useMemo(() => {
    const unconfTxList = txHelper(
      unapprovedTxs || {},
      unapprovedPersonalMsgs,
      {},
      {},
      unapprovedTypedMessages,
      chainId,
    );

    log.info(`rendering a combined ${unconfTxList.length} unconf msgs & txs`);

    const unconfirmedTx = txIdFromPath
      ? unconfTxList.find(({ id }) => `${id}` === txIdFromPath)
      : unconfTxList[index];
    return unconfirmedTx ? cloneDeep(unconfirmedTx) : {};
  }, [
    chainId,
    index,
    txIdFromPath,
    unapprovedPersonalMsgs,
    unapprovedTxs,
    unapprovedTypedMessages,
  ]);

  const { warnings } = useInsightSnaps(txData.id);
  const resolvedSecurityAlertResponse =
    signatureSecurityAlertResponses?.[
      txData.securityAlertResponse?.securityAlertId
    ];

  if (resolvedSecurityAlertResponse) {
    txData.securityAlertResponse = resolvedSecurityAlertResponse;
  }

  const targetSubjectMetadata = useSelector((state) =>
    getTargetSubjectMetadata(state, txData.msgParams?.origin),
  );

  if (!txData.msgParams) {
    return <Loading />;
  }

  const SigComponent = signatureSelect(txData, targetSubjectMetadata);

  return (
    <SigComponent
      history={history}
      txData={txData}
      key={txData.id}
      accounts={internalAccounts}
      currentCurrency={currentCurrency}
      blockGasLimit={blockGasLimit}
      warnings={warnings}
    />
  );
};

ConfirmTxScreen.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
    }),
  }),
};

export default withRouter(ConfirmTxScreen);
