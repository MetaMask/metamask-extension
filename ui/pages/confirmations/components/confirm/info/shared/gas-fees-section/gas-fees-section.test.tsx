import { act } from '@testing-library/react';
import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import {
  getMockConfirmState,
  getMockContractInteractionConfirmState,
} from '../../../../../../../../test/data/confirmations/helper';
import { renderWithConfirmContextProvider } from '../../../../../../../../test/lib/confirmations/render-helpers';
import { getGasFeeTimeEstimate } from '../../../../../../../store/actions';
import { GasFeesSection } from './gas-fees-section';

jest.mock('../../../../../../../store/actions', () => ({
  ...jest.requireActual('../../../../../../../store/actions'),
  getGasFeeTimeEstimate: jest.fn(),
}));

jest.mock(
  '../../../../../../../components/app/alert-system/contexts/alertMetricsContext',
  () => ({
    useAlertMetrics: jest.fn(() => ({
      trackAlertMetrics: jest.fn(),
    })),
  }),
);

describe('<GasFeesSection />', () => {
  const middleware = [thunk];

  it('does not render component for gas fees section', () => {
    const state = getMockConfirmState();
    const mockStore = configureMockStore(middleware)(state);
    const { container } = renderWithConfirmContextProvider(
      <GasFeesSection />,
      mockStore,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders component for gas fees section', async () => {
    (getGasFeeTimeEstimate as jest.Mock).mockImplementation(() =>
      Promise.resolve({ upperTimeBound: '1000' }),
    );

    const state = getMockContractInteractionConfirmState();
    const mockStore = configureMockStore(middleware)(state);
    let container;
    await act(async () => {
      const renderResult = renderWithConfirmContextProvider(
        <GasFeesSection />,
        mockStore,
      );
      container = renderResult.container;

      // Wait for any asynchronous operations to complete
      await new Promise(setImmediate);
    });

    expect(container).toMatchSnapshot();
  });
});
