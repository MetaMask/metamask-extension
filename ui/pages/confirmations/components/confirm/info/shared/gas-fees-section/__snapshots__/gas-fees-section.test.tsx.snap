// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GasFeesSection /> does not render component for gas fees section 1`] = `<div />`;

exports[`<GasFeesSection /> renders component for gas fees section 1`] = `
<div>
  <div
    class="mm-box mm-box--margin-bottom-4 mm-box--padding-2 mm-box--background-color-background-default mm-box--rounded-md"
    data-testid="gas-fee-section"
  >
    <div
      class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
      data-testid="edit-gas-fees-row"
      style="overflow-wrap: anywhere; min-height: 24px; position: relative; background: transparent;"
    >
      <div
        class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start mm-box--color-text-default"
      >
        <div
          class="mm-box mm-box--display-flex mm-box--align-items-center"
        >
          <p
            class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
          >
            Network fee
          </p>
          <div>
            <div
              aria-describedby="tippy-tooltip-1"
              class=""
              data-original-title="Amount paid to process the transaction on network."
              data-tooltipped=""
              style="display: flex;"
              tabindex="0"
            >
              <span
                class="mm-box mm-icon mm-icon--size-sm mm-box--margin-left-1 mm-box--display-inline-block mm-box--color-icon-muted"
                data-testid="edit-gas-fees-row-tooltip"
                style="mask-image: url('./images/icons/question.svg');"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-space-between mm-box--align-items-center mm-box--text-align-center"
      >
        <p
          class="mm-box mm-text mm-text--body-md mm-box--margin-right-1 mm-box--color-text-default"
          data-testid="first-gas-field"
        >
          0.0001 ETH
        </p>
        <p
          class="mm-box mm-text mm-text--body-md mm-box--margin-right-2 mm-box--color-text-alternative"
          data-testid="native-currency"
        >
          $0.04
        </p>
        <button
          class="mm-box mm-text mm-button-base mm-button-base--size-sm mm-button-link mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent"
          data-testid="edit-gas-fee-icon"
          style="text-decoration: none;"
        >
          <span
            class="mm-box mm-icon mm-icon--size-sm mm-box--margin-inline-end-1 mm-box--display-inline-block mm-box--color-inherit"
            style="mask-image: url('./images/icons/edit.svg');"
          />
          <span
            class="mm-box mm-text mm-text--inherit mm-box--color-primary-default"
          />
        </button>
      </div>
    </div>
    <div
      class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
      data-testid="gas-fee-details-speed"
      style="overflow-wrap: anywhere; min-height: 24px; position: relative; background: transparent;"
    >
      <div
        class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start mm-box--color-text-default"
      >
        <div
          class="mm-box mm-box--display-flex mm-box--align-items-center"
        >
          <p
            class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
          >
            Speed
          </p>
        </div>
      </div>
      <div
        class="mm-box mm-box--display-flex mm-box--align-items-center"
      >
        <div
          class="mm-box mm-box--display-flex mm-box--flex-wrap-wrap"
        >
          <p
            class="mm-box mm-text mm-text--body-md mm-box--padding-inline-end-1 mm-box--color-text-alternative"
          >
            ðŸ¦Š Market
          </p>
          <p
            class="mm-box mm-text mm-text--body-md mm-box--color-text-default"
          >
            <span
              data-testid="gas-timing-time"
            >
              ~
              0 sec
            </span>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
